src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -123,6 +123,28 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
                 result = SetResolveData(result.Value, serverKind);
             }
 
+            // This is a bit fraught. Basically in the scenario That someone types "@<slight pause>{" they'll get @__builder because '{' is a commit character.
+            // Our solution is to make the results in this case empty, which is a bit of a lie.
+            // It's the best we can do for now, if a user complains we can revisit.
+            if (result.HasValue && request.Context.TriggerCharacter != null && request.Context.TriggerCharacter.Equals("@"))
+            {
+                if (result.Value.Value is CompletionItem[] items)
+                {
+                    foreach (var item in items)
+                    {
+                        item.InsertText = string.Empty;
+                    }
+                }
+                else
+                {
+                    var list = result.Value.Value as CompletionList;
+                    foreach (var item in list?.Items)
+                    {
+                        item.InsertText = string.Empty;
+                    }
+                }
```

Comment:
Can you clarify this a bit for me? We're letting the completion entries stay, but just modifying their insertion text to nothing? Why even show them?

Reply:
This was my initial reaction too, but other solutions seemed to end up with worse experiences and @NTaylorMullen agreed.

- Just removing '{' as a commit character ended up with the completion dialog not dismissing and you ending up with @__builder when you hit enter after your {.
- Removing all the results in this class results in ending up with only completions for `@functions` and `@code`, which then suffer the same problems as `@__builder`, but are much more likely to be the result that people actually wanted.

This does end up with empty commits for `@` completions, but even one additional character avoids that scenario. @NTaylorMullen to make sure he still agrees with this solution when it's put in full context.

Reply:
An alternative solution here would be for VS to re-invoke or dismiss completion in this case. [We had a similar](https://github.com/dotnet/aspnetcore-tooling/blob/master/src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCompletionItemProvider.ts#L105) problem in VSCode and resolved it by removing "{" as a commit character in this case. Since VSCode re-invoked and/or cleared completions when that happened everything afterward worked.

Reply:
> Why even show them?

More of an experience piece. Users are used to seeing a default C# completion list pop-up once `@` is typed to aid in their further typing of completions for filtering concerns.